python/ray/tests/test_reference_counting.py

```diff
@@ -236,15 +258,296 @@ def get_large_object(self):
+@pytest.mark.skip("Serialized ObjectID reference counting not implemented.")
+def test_actor_holding_serialized_reference(one_worker_100MiB):
+    @ray.remote
+    class GreedyActor(object):
+        def __init__(self):
+            pass
+
+        def set_ref1(self, ref):
+            self.ref1 = ref
+
+        def add_ref2(self, new_ref):
+            self.ref2 = new_ref
+
+        def delete_ref1(self):
+            self.ref1 = None
+
+        def delete_ref2(self):
+            self.ref2 = None
+
+    @ray.remote
+    def put():
+        return np.zeros(40 * 1024 * 1024, dtype=np.uint8)
+
+    # Test that the reference held by the actor isn't evicted.
+    array_oid = put.remote()
+    actor = GreedyActor.remote()
+    actor.set_ref1.remote([array_oid])
+
+    # Remove the local reference.
+    array_oid_bytes = array_oid.binary()
+    del array_oid
+
+    # Test that the remote reference still pins the object.
+    _fill_object_store_and_get(array_oid_bytes)
+
+    # Test that giving the same actor a duplicate reference works.
+    ray.get(actor.add_ref2.remote([ray.ObjectID(array_oid_bytes)]))
+    _fill_object_store_and_get(array_oid_bytes)
+
+    # Test that removing only the first reference doesn't unpin the object.
+    ray.get(actor.delete_ref1.remote())
+    _fill_object_store_and_get(array_oid_bytes)
+
+    # Test that deleting the second reference stops it from being pinned.
+    ray.get(actor.delete_ref2.remote())
+    _fill_object_store_and_get(array_oid_bytes, succeed=False)
+
+
+# Test that a passed reference held by an actor after a task finishes
+# is kept until the reference is removed from the worker. Also tests giving
+# the worker a duplicate reference to the same object ID.
+@pytest.mark.skip("Serialized ObjectID reference counting not implemented.")
+def test_worker_holding_serialized_reference(one_worker_100MiB):
+    @ray.remote
+    def child(dep1, dep2):
+        return
+
+    @ray.remote
+    def launch_pending_task(refs):
+        ref, dep = refs
+        return child.remote(ref, dep)
+
+    @ray.remote
+    def put():
+        return np.zeros(40 * 1024 * 1024, dtype=np.uint8)
+
+    # Test that the reference held by the actor isn't evicted.
+    array_oid = put.remote()
+    random_oid = ray.ObjectID.from_random()
+    child_return_id = ray.get(
+        launch_pending_task.remote([array_oid, random_oid]))
+
+    # Remove the local reference.
+    array_oid_bytes = array_oid.binary()
+    del array_oid
+
+    # Test that the reference prevents the object from being evicted.
+    _fill_object_store_and_get(array_oid_bytes)
+
+    ray.worker.global_worker.put_object(None, object_id=random_oid)
+    ray.get(child_return_id)
+    del child_return_id
+
+    _fill_object_store_and_get(array_oid_bytes, succeed=False)
+
+
+# Test that an object containing object IDs within it pins the inner IDs.
+def test_basic_nested_ids(one_worker_100MiB):
+    inner_oid = ray.put(np.zeros(40 * 1024 * 1024, dtype=np.uint8))
```

Comment:
This test looks good, but we should probably also add one that passes nested IDs to other tasks. So something that nests an ID to some depth, then submits a bunch of tasks recursively that each unwrap the ID once.

Reply:
added